cloud/linode/route_controller.go

```diff
@@ -0,0 +1,237 @@
+package linode
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/linode/linodego"
+	"golang.org/x/exp/slices"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+)
+
+type routeCache struct {
+	sync.RWMutex
+	routes     map[int][]linodego.InstanceConfig
+	lastUpdate time.Time
+	ttl        time.Duration
+}
+
+func (rc *routeCache) refreshRoutes(ctx context.Context, client Client) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if time.Since(rc.lastUpdate) < rc.ttl {
+		return nil
+	}
+
+	instances, err := client.ListInstances(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	rc.routes = make(map[int][]linodego.InstanceConfig, len(instances))
+	wg := sync.WaitGroup{}
+	for _, instance := range instances {
+		wg.Add(1)
+		go func(ctx context.Context, id int) {
+			configs, err := client.ListInstanceConfigs(ctx, id, &linodego.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed fetching instance configs for instance id %d. Error: %s", id, err.Error())
+			}
+			rc.routes[id] = configs
+			wg.Done()
+		}(ctx, instance.ID)
+	}
+
+	wg.Wait()
+
+	rc.lastUpdate = time.Now()
+	return nil
+}
+
+type routes struct {
+	vpcid      int
+	client     Client
+	instances  *instances
+	routeCache *routeCache
+}
+
+func newRoutes(client Client) (cloudprovider.Routes, error) {
+	timeout := 60
+	if raw, ok := os.LookupEnv("LINODE_ROUTES_CACHE_TTL"); ok {
+		if t, _ := strconv.Atoi(raw); t > 0 {
+			timeout = t
+		}
+	}
+	klog.V(3).Infof("TTL for routeCache set to %d", timeout)
+
+	vpcid, err := getVPCID(client, Options.VPCName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &routes{
+		vpcid:     vpcid,
+		client:    client,
+		instances: newInstances(client),
+		routeCache: &routeCache{
+			routes: make(map[int][]linodego.InstanceConfig),
```

Comment:
```suggestion
			routes: make(map[int][]linodego.InstanceConfig, 0),
```

The map here will be thrown away the first time `refreshRoutes` is called, so I suggest you set the capacity to 0

Reply:
Updated.